src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -159,17 +181,24 @@ class IndexAllocation implements NativeMemoryAllocation {
             this.watcherHandle = watcherHandle;
             this.sharedIndexState = sharedIndexState;
             this.isBinaryIndex = isBinaryIndex;
+            this.refCounted = new RefCountedReleasable<>("IndexAllocation-Reference", this, this::closeInternal);
         }
 
-        @Override
-        public void close() {
+        public void closeInternal() {
             executor.execute(() -> {
                 writeLock();
                 cleanup();
                 writeUnlock();
             });
         }
 
+        @Override
+        public void close() {
+            if (!closed && refCounted.refCount() > 0) {
```

Comment:
Wait if refCount is > 0, we dont want to close right?

Reply:
in concurrent scenarios, there is an search happen before the close. and we do can not assume `KNNWeight#doANNSearch` or `IndexShard#Close` or `ClearCache` has already returned like async close shard/segment scenarios. so i prefer to keep the memory incase `SegmentFault`.